pilot/pkg/model/destination_rule.go

```diff
@@ -38,55 +39,76 @@ import (
+		// `addRuleToProcessedDestRules` determines if the incoming destination rule would become a new unique entry in the processedDestRules list.
+		addRuleToProcessedDestRules := true
+		for i, mdr := range mdrList {
+			// Deep copy destination rule, to prevent mutate it later when merge with a new one.
+			// This can happen when there are more than one destination rule of same host in one namespace.
+			copied := mdr.DeepCopy()
+			p.destRules[resolvedHost][i] = &copied
+			mergedRule := copied.Spec.(*networking.DestinationRule)
+			bothWithoutSelector := rule.GetWorkloadSelector() == nil && mergedRule.GetWorkloadSelector() == nil
+			bothWithSelector := mergedRule.GetWorkloadSelector() != nil && rule.GetWorkloadSelector() != nil
+			selectorsMatch := labels.Instance(mergedRule.GetWorkloadSelector().GetMatchLabels()).Equals(rule.GetWorkloadSelector().GetMatchLabels())
 
-	if mdr, exists := p.destRule[resolvedHost]; exists {
-		// Deep copy destination rule, to prevent mutate it later when merge with a new one.
-		// This can happen when there are more than one destination rule of same host in one namespace.
-		copied := mdr.DeepCopy()
-		p.destRule[resolvedHost] = &copied
-		mergedRule := copied.Spec.(*networking.DestinationRule)
-		existingSubset := map[string]struct{}{}
-		for _, subset := range mergedRule.Subsets {
-			existingSubset[subset.Name] = struct{}{}
-		}
-		// we have an another destination rule for same host.
-		// concatenate both of them -- essentially add subsets from one to other.
-		// Note: we only add the subsets and do not overwrite anything else like exportTo or top level
-		// traffic policies if they already exist
-		for _, subset := range rule.Subsets {
-			if _, ok := existingSubset[subset.Name]; !ok {
-				// if not duplicated, append
-				mergedRule.Subsets = append(mergedRule.Subsets, subset)
-			} else {
-				// duplicate subset
-				ps.AddMetric(DuplicatedSubsets, string(resolvedHost), "",
-					fmt.Sprintf("Duplicate subset %s found while merging destination rules for %s",
-						subset.Name, string(resolvedHost)))
+			if bothWithSelector && !selectorsMatch {
+				// If the new destination rule and the existing one has workload selectors associated with them, skip merging
+				// if the selectors do not match
+				continue
 			}
-		}
 
-		// If there is no top level policy and the incoming rule has top level
-		// traffic policy, use the one from the incoming rule.
-		if mergedRule.TrafficPolicy == nil && rule.TrafficPolicy != nil {
-			mergedRule.TrafficPolicy = rule.TrafficPolicy
-		}
+			// If both the destination rules are without a workload selector or with matching workload selectors, simply merge them.
+			// If the incoming rule has a workload selector, it has to be merged with the existing rules with workload selector, and
+			// at the same time added as a unique entry in the processedDestRules.
+			if bothWithoutSelector || (rule.GetWorkloadSelector() != nil && selectorsMatch) {
```

Comment:
have you checked when one has selector, but the other not.

Reply:
@kfaseela Can you point me?  Where do they merge?

Reply:
So come back to original doubt https://github.com/istio/istio/pull/37174#discussion_r832791802

Reply:
if one has selector, and the other one does not have selector, we will have "addRuleToProcessedDestRules=true". And as the comment here says, in such cases, rule with selector merges to the rule without selector, as well as at line [102](https://github.com/istio/istio/pull/37174/files#diff-3320c5cd8bd7f37eca4dead440d5f0999739bd17c5939e4e2340f82a33f3f835R102), we add the new entry separately to the list. Is it clear now @hzxuzhonghu  ?

Reply:
yep